x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJob.java

```diff
@@ -217,44 +231,269 @@ public ConnectorSyncJob(StreamInput in) throws IOException {
         this.workerHostname = in.readOptionalString();
     }
 
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<ConnectorSyncJob, Void> PARSER = new ConstructingObjectParser<>(
+        "connector_sync_job",
+        true,
+        (args) -> {
+            int i = 0;
+            return new Builder().setCancellationRequestedAt((Instant) args[i++])
+                .setCanceledAt((Instant) args[i++])
+                .setCompletedAt((Instant) args[i++])
+                .setConnector((Connector) args[i++])
+                .setCreatedAt((Instant) args[i++])
+                .setDeletedDocumentCount((Long) args[i++])
+                .setError((String) args[i++])
+                .setId((String) args[i++])
+                .setIndexedDocumentCount((Long) args[i++])
+                .setIndexedDocumentVolume((Long) args[i++])
+                .setJobType((ConnectorSyncJobType) args[i++])
+                .setLastSeen((Instant) args[i++])
+                .setMetadata((Map<String, Object>) args[i++])
+                .setStartedAt((Instant) args[i++])
+                .setStatus((ConnectorSyncStatus) args[i++])
+                .setTotalDocumentCount((Long) args[i++])
+                .setTriggerMethod((ConnectorSyncJobTriggerMethod) args[i++])
+                .setWorkerHostname((String) args[i])
+                .build();
+        }
+    );
+
+    static {
+        PARSER.declareField(
+            optionalConstructorArg(),
+            (p, c) -> Instant.parse(p.text()),
+            CANCELATION_REQUESTED_AT_FIELD,
+            ObjectParser.ValueType.STRING
+        );
+        PARSER.declareField(optionalConstructorArg(), (p, c) -> Instant.parse(p.text()), CANCELED_AT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField(optionalConstructorArg(), (p, c) -> Instant.parse(p.text()), COMPLETED_AT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField(
+            constructorArg(),
+            (p, c) -> ConnectorSyncJob.syncJobConnectorFromXContent(p),
+            CONNECTOR_FIELD,
+            ObjectParser.ValueType.OBJECT
+        );
+        PARSER.declareField(constructorArg(), (p, c) -> Instant.parse(p.text()), CREATED_AT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareLong(constructorArg(), DELETED_DOCUMENT_COUNT_FIELD);
+        PARSER.declareString(optionalConstructorArg(), ERROR_FIELD);
```

Comment:
I see you followed the approach to deserialize index data into object before returning the data to user (it's a convenient way if we want to support secret obfuscation in the future :+1: ). 

One thing to be careful about here is handling `null` values. For example in status quo I see that error field can have value set either to `null` or `string`. Error field parsing only supports string. Therefore if you would try to retrieve data sync job created e.g. in 8.11 release with this API it would fail as the property `"error": null,` is present in the index. I think we need to use `declareStringOrNull` here - to make it backward compatible. 

I'm aware that this issue might be also present for certain fields in Connector class, I intend to go over this in the QA phase. It's up to you if you want to dig into this now or adjust it in the followup ticket (it's only about tweaking PARSER, toXContent and adding few testcases)

Reply:
Addressed with https://github.com/elastic/elasticsearch/pull/102908/commits/c22377d8d7d95d33d11f75629f6cf6be6770d59c